opendevin/controller/agent_controller.py

```diff
@@ -144,17 +148,17 @@ async def step(self, i: int):
         try:
             action = self.agent.step(self.state)
             if action is None:
-                raise ValueError('Agent must return an action')
+                raise AgentNoActionError()
             print_with_color(action, 'ACTION')
         except Exception as e:
             observation = AgentErrorObservation(str(e))
             print_with_color(observation, 'ERROR')
             traceback.print_exc()
-            # TODO Change to more robust error handling
-            if (
-                'The api_key client option must be set' in observation.content
-                or 'Incorrect API key provided:' in observation.content
-            ):
+            if isinstance(e, APIConnectionError):
+                time.sleep(3)
+
+            # raise specific exceptions that need to be handled outside
+            if isinstance(e, (AuthenticationError, AgentNoActionError)):
```

Comment:
> For me, on openai, the auth error from those is an openai.AuthenticationError

That's correct but litellm then catches it and then wraps it with AuthenticationError

You could try this out locally by providing a false or empty OpenAPI key. Unfortunately we don't have proper infra to write a test for this yet.

Reply:
> I think the first LLM calls that happen are embeddings calls. From what I see, those are not handled via litellm, but via llama-index. (memory.py)

Oh I see what you mean. You are saying this exception handling won't catch embedding calls' exceptions. That might very likely be true. Let me play around with this.

Reply:
Interesting, I did, and it didn't work for me. Litellm isn't sending those calls. I wonder how it did for you 😅 😕

Reply:
Interesting, let me see if I can make a screenshot

Reply:
My config:

<img width="256" alt="image" src="https://github.com/OpenDevin/OpenDevin/assets/25746010/63c5a895-f15d-4049-9e1e-d3104f402674">

Demo (GUI):

<img width="481" alt="image" src="https://github.com/OpenDevin/OpenDevin/assets/25746010/1f572623-7a88-415a-8944-dced8407043d">

Demo (Backend):

(Note: I added two adhoc lines for demo purpose)

<img width="1110" alt="Screenshot 2024-04-13 at 8 45 54 PM" src="https://github.com/OpenDevin/OpenDevin/assets/25746010/1e720d91-1a96-48f0-a7cb-b3d0b3cc9293">


<img width="813" alt="image" src="https://github.com/OpenDevin/OpenDevin/assets/25746010/15ba13b5-5975-4113-a83b-0ca2bd0da1e5">

I believe the above screenshots suffice to prove that this works on MY MACHINE. Now let's prove this conversely - if it doesn't work, what shall happen?

You will see the same error message keeps popping up because the exception is not properly caught and raised.

(Note: I removed the `AuthenticationError` from `isinstance` check for demo purpose)

<img width="465" alt="image" src="https://github.com/OpenDevin/OpenDevin/assets/25746010/0bfe296f-5c64-4537-bdcb-883b71d1fd81">

Reply:
@enyst Just to make sure we are on the same page - when you say it didn't work for you, did you mean by embeddings call? If yes then you might be true - I didn't verify that.

Reply:
I got it, I see what's happening.

New installation, with variables set, but the API_KEY looks right and it's wrong.
https://gist.github.com/enyst/4961b7b60d1f6915336e6fcb45ea62a3
This goes through llama-index and it's an openai exception.

New installation, with no variables set because I forgot, and started it with no key, no model:
This is where I see the litellm completion call failing first, instead. I'll have to repeat to get the traceback but it's probably what you're seeing.

Reply:
Yes, I mean the embeddings call. If that goes off first, it fails, and it fails without litellm in the process. Finally I understand why we had the first string, hah! The second string is from the embeddings failure, but the first is from completion failure. I usually see embeddings failing first. 😆

Reply:
Yeah thanks for catching this. I didn't meet this problem because I didn't use OpenAI embedding as part of my testing. Now I can reproduce what you found and indeed this PR brings a REGRESSION!

Reply:
Fixed. Now:

1) AuthError from embedding call to OpenAI is raised and task is terminated (so far, OpenAI is the only embedding provider we support that needs auth, but I didn't try Azure OpenAI).
2) AuthError from completion call to any LLM is raised and task is terminated